canal/handler.go

```diff
@@ -17,6 +17,14 @@ type EventHandler interface {
 	OnGTID(header *replication.EventHeader, gtid mysql.GTIDSet) error
 	// OnPosSynced Use your own way to sync position. When force is true, sync position immediately.
 	OnPosSynced(header *replication.EventHeader, pos mysql.Position, set mysql.GTIDSet, force bool) error
+	// OnCreateUser is the create user event in QueryEvent
```

Comment:
My concern is, the rest methods of this interface are related to binlog event type, not DDL type. So what do you think we add another interface `QueryEventHandler` to handle it? BTW, please add comment to describe the callback order of `OnDDL` and these newly added methods

Reply:
Make the canal better meet the synchronization of two heterogeneous or remote data centers, usually the two data centers contain the synchronization of data and user rights

---

I removed `QueryEventHandler ` define, and changed to a more understanding case mode

```go
// handleQueryEvent is handle some common query events (e.g., DDL,CREATE or DROP USER,GRANT),
// others use UnknownQueryEvent unified callbacks to expose to users
func (c *Canal) handleQueryEvent(ev *replication.BinlogEvent, e *replication.QueryEvent,
	stmt ast.StmtNode, pos mysql.Position, savePos, force *bool) error {
	switch t := stmt.(type) {
	case *ast.RenameTableStmt, *ast.AlterTableStmt, *ast.DropTableStmt, *ast.CreateTableStmt, *ast.TruncateTableStmt:
		return c.handleDDLEvent(ev, e, t, pos, savePos, force)
	case *ast.CreateUserStmt:
		return c.handleCreateUserEvent(ev, e, t, pos, savePos, force)
	case *ast.DropUserStmt:
		return c.handleDropUserEvent(ev, e, t, pos, savePos, force)
	case *ast.GrantStmt:
		return c.handleGrantEvent(ev, e, t, pos, savePos, force)
	default:
		return c.handleUnknownQueryEvent(ev, e, t, pos, savePos, force)
	}
}
```

Reply:
> the rest methods of this interface are related to binlog event type, not DDL type

Please explain how your solution solves this confusing point.

> add comment to describe the callback order of OnDDL and these newly added methods

And this one

Reply:
> the rest methods of this interface are related to binlog event type, not DDL type

sorry, The type `QUERY_EVENT` in the `binlog` log in mysq contains not only `DDL` operations, but also changes to the database,     such as：`Create User`

In our `github.com/go-mysql-org/go-mysql`, only DDL Type is parsed in the Canal package，and now some event parsing has been added

> add comment to describe the callback order of OnDDL and these newly added methods

Sorry, I had a problem with my previous implementation, they have no dependencies, they all belong to the subtypes of `QUERY_EVENT`  and changed to a more understanding case mode

---
```go
func (c *Canal) runSyncBinlog() error {
   //.....
   for {
   //.....
         switch e := ev.Event.(type) {
		//.....
                case *replication.QueryEvent:
			stmts, _, err := c.parser.Parse(string(e.Query), "", "")
			if err != nil {
				c.cfg.Logger.Errorf("parse query(%s) err %v, will skip this event", e.Query, err)
				continue
			}
			for _, stmt := range stmts {
				err := c.handleQueryEvent(ev, e, stmt, pos, &savePos, &force)
				if err != nil {
					c.cfg.Logger.Errorf("handle query event(%s) err %v", e.Query, err)
				}
			}
               //.......
      }
   //.....

// handleQueryEvent is handle some common query events (e.g., DDL,CREATE or DROP USER,GRANT),
// others use UnknownQueryEvent unified callbacks to expose to users
func (c *Canal) handleQueryEvent(ev *replication.BinlogEvent, e *replication.QueryEvent,
	stmt ast.StmtNode, pos mysql.Position, savePos, force *bool) error {
	switch t := stmt.(type) {
	case *ast.RenameTableStmt, *ast.AlterTableStmt, *ast.DropTableStmt, *ast.CreateTableStmt, *ast.TruncateTableStmt:
		return c.handleDDLEvent(ev, e, t, pos, savePos, force)
	case *ast.CreateUserStmt:
		return c.handleCreateUserEvent(ev, e, t, pos, savePos, force)
	case *ast.DropUserStmt:
		return c.handleDropUserEvent(ev, e, t, pos, savePos, force)
	case *ast.GrantStmt:
		return c.handleGrantEvent(ev, e, t, pos, savePos, force)
	default:
		return c.handleUnknownQueryEvent(ev, e, t, pos, savePos, force)
	}
}
```

Reply:
Oh I understand, thanks for explanation

Reply:
i think it also contains the begin of a transcation with the word BEGIN, will a EventHandler.OnTransBegin here?